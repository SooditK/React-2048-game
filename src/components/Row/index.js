import React from 'react';
import PropTypes from 'prop-types';
import Cell from '../Cell';
import { isObjEqual } from '../../utils/helpers';

// Game board row, contain 4 cell
export default class Row extends React.Component {
  static propTypes = {
    row: PropTypes.arrayOf(PropTypes.number).isRequired
  };

  shouldComponentUpdate(nextProps, nextState) {
    return (
      !isObjEqual(nextProps, this.props) || !isObjEqual(nextState, this.state)
    );
  }

  render() {
    const {
      props: { row }
    } = this;

    return (
      <tr>
        {/* eslint-disable react/no-array-index-key */}
        {row.map((num, idx) => (
          <Cell value={num} key={idx} />
        ))}
      </tr>
    );
  }
}
