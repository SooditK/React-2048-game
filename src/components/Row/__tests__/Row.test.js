import React from 'react';
import renderer from 'react-test-renderer';
import Enzyme, { mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import Row from '..';

Enzyme.configure({ adapter: new Adapter() });

describe('<Board />', () => {
  it('component render', () => {
    const MATRIX = [[2, 0, 0, 0], [0, 0, 0, 0], [0, 0, 4, 0], [0, 0, 0, 0]];
    MATRIX.forEach(r => {
      const row = renderer.create(<Row row={r} />).toJSON();
      expect(row).toMatchSnapshot();
    });
  });

  it('shouldComponentUpdate', () => {
    const data = [2, 0, 0, 0];
    const tbody = document.createElement('tbody');
    const cell = mount(<Row row={data} />, {
      attachTo: tbody
    });
    expect(cell.props().row).toEqual(data);
    let shouldUpdate = cell.instance().shouldComponentUpdate(
      {
        row: data
      },
      null
    );
    expect(shouldUpdate).toBe(false);
    shouldUpdate = cell.instance().shouldComponentUpdate(
      {
        value: [2, 4, 8, 16]
      },
      null
    );
    expect(shouldUpdate).toBe(true);
  });
});
