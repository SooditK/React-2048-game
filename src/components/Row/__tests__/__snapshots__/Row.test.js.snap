// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Board /> component render 1`] = `
<tr>
  <td>
    <div
      className="cell color-2"
    >
      <div
        className="number"
      >
        2
      </div>
    </div>
  </td>
  <td>
    <div
      className="cell"
    >
      <div
        className="number"
      />
    </div>
  </td>
  <td>
    <div
      className="cell"
    >
      <div
        className="number"
      />
    </div>
  </td>
  <td>
    <div
      className="cell"
    >
      <div
        className="number"
      />
    </div>
  </td>
</tr>
`;

exports[`<Board /> component render 2`] = `
<tr>
  <td>
    <div
      className="cell"
    >
      <div
        className="number"
      />
    </div>
  </td>
  <td>
    <div
      className="cell"
    >
      <div
        className="number"
      />
    </div>
  </td>
  <td>
    <div
      className="cell"
    >
      <div
        className="number"
      />
    </div>
  </td>
  <td>
    <div
      className="cell"
    >
      <div
        className="number"
      />
    </div>
  </td>
</tr>
`;

exports[`<Board /> component render 3`] = `
<tr>
  <td>
    <div
      className="cell"
    >
      <div
        className="number"
      />
    </div>
  </td>
  <td>
    <div
      className="cell"
    >
      <div
        className="number"
      />
    </div>
  </td>
  <td>
    <div
      className="cell color-4"
    >
      <div
        className="number"
      >
        4
      </div>
    </div>
  </td>
  <td>
    <div
      className="cell"
    >
      <div
        className="number"
      />
    </div>
  </td>
</tr>
`;

exports[`<Board /> component render 4`] = `
<tr>
  <td>
    <div
      className="cell"
    >
      <div
        className="number"
      />
    </div>
  </td>
  <td>
    <div
      className="cell"
    >
      <div
        className="number"
      />
    </div>
  </td>
  <td>
    <div
      className="cell"
    >
      <div
        className="number"
      />
    </div>
  </td>
  <td>
    <div
      className="cell"
    >
      <div
        className="number"
      />
    </div>
  </td>
</tr>
`;
